be/src/service/backend_service.h

```diff
@@ -148,11 +166,32 @@ class BackendService : public BackendServiceIf {
     virtual void get_tablet_stat(TTabletStatResult& result) override;
 
     virtual void submit_routine_load_task(TStatus& t_status, const std::vector<TRoutineLoadTask>& tasks) override;
+
+    // used for external service, open means start the scan procedure
+    virtual void open_scanner(TScanOpenResult& result_, const TScanOpenParams& params);
+
+    // used for external service, external use getNext to fetch data batch after batch until eos = true
+    virtual void get_next(TScanBatchResult& result_, const TScanNextBatchParams& params);
+
+    // used for external service, close some context and release resource related with this context
+    virtual void close_scanner(TScanCloseResult& result_, const TScanCloseParams& params);
+
 private:
     Status start_plan_fragment_execution(const TExecPlanFragmentParams& exec_params);
 
     ExecEnv* _exec_env;
     std::unique_ptr<AgentServer> _agent_server;
+
+    std::map<std::string, std::shared_ptr<Context>> _active_contexts;
```

Comment:
don't put this in backend service, I think you should create a class like ExternalScannerManger to handle these things